A position-specific weight matrix trait with implementations for both DNA and protein sequences

hervold opened the pull request:
PSSM's (aka PWM's) are pretty handy, and serve as the basis of tools like Pfam and Psi-BLAST.
I believe my implementation is pretty straight-forward, but I'd welcome any criticism.
Thanks for taking a look!

johanneskoester commented on src/pattern_matching/pssm/mod.rs:
```diff
@@ -0,0 +1,154 @@
+// except according to those terms.
+
+//! Implementation of position-specific scoring matrix (PSSM), aka position
+//! weight matrix (PWM), for both DNA and amino-acid sequences
```
Could you add an example usage here?

hervold replied:
How do those examples and doc strings look?
I also cleaned up error-handling a bit -- eg, an erroneous base (eg, an N) used to cause a panic!, but now just returns Err

johanneskoester commented on src/pattern_matching/pssm/mod.rs:
```diff
@@ -0,0 +1,154 @@
+        (best_start, best_score, best_m)
+    }
+
+    /// apply PSM to sequence, finding the offset with the highest score
```
This should be PSSM, not PSM? In general, could this PR you just one name, either PSSM or PWM? Or is it necessay to distinguish?

johanneskoester commented on src/pattern_matching/pssm/mod.rs:
```diff
@@ -0,0 +1,209 @@
+
+/// Errors
+#[derive(Debug, Clone)]
+pub enum PSSMError {
```
Please use quick_error macro for this, see https://github.com/rust-bio/rust-bio/blob/66c7de44aae1c28780f1245e959a8127dd1f63ae/src/utils/interval.rs#L55

hervold commented:
@johanneskoester - I added an explanation of the algorithm along the lines you suggested; does that look ok?  I've also added some tests for the error-handling.

johanneskoester commented on src/pattern_matching/pssm/dnamotif.rs:
```diff
@@ -0,0 +1,269 @@
+    }
+
+    // helper function
+    pub fn calc_minmax(&mut self) {
```
if this is a helper function, I think it should be private. Alternatively, please add a doc comment.

johanneskoester commented on src/pattern_matching/pssm/dnamotif.rs:
```diff
@@ -0,0 +1,269 @@
+}
+
+impl DNAMotif {
+    pub fn from_seqs_with_pseudocts(
```
Please add a doc comment.

johanneskoester commented on src/pattern_matching/pssm/dnamotif.rs:
```diff
@@ -0,0 +1,269 @@
+
+    /// derived from
+    /// https://github.com/biopython/biopython/blob/master/Bio/motifs/matrix.py#L205
+    fn degenerate_consensus(&self) -> Result<Vec<u8>, PSSMError> {
```
Change this to a normal comment inside the function. Otherwise, it would overwrite the doc comment of the trait.

johanneskoester commented on src/pattern_matching/pssm/mod.rs:
```diff
@@ -0,0 +1,229 @@
+
+    /// Standard PSSM scoring is calibrated to (1) motif len and (2) the min and
+    /// max possible scores.  This is just a un-normalized sum of matching bases, useful for
+    /// comparing matches from motifs of different lengths.
```
Please explain args and return values in all doc comments.

hervold commented:
@johanneskoester  - I think I've added doc comments to every public entity now.

Thanks for bearing with me - I've never contributed anything significant to an open source project before, so my standards for documentation are a little weak.  I really appreciate your taking the time to look at this!

johanneskoester commented on src/pattern_matching/pssm/dnamotif.rs:
```diff
@@ -0,0 +1,244 @@
+    /// # Arguments
+    /// * `seqs` - sequences incorportated into motif
+    /// * `pseudos` - array slice with a pseudocount for each monomer;
+    ///    defaults to DEF_PSEUDO for all if None is supplied
```
Say ```pssm::DEF_PSEUDO``` here and make it public. Otherwise, nobody would be able to see what is the default.

johanneskoester commented:
Great! May I ask you to format with Rust 1.26.2, so that travis CI will be happy?

johanneskoester commented:
Thanks!
